File: src/main/java/pt/haslab/dql/tictactoe/agents/MrMiyagiAgent.java
package pt.haslab.dql.tictactoe.agents;

import pt.haslab.dql.tictactoe.game.Actions;
import pt.haslab.dql.tictactoe.game.Board;
import pt.haslab.dql.tictactoe.game.Seed;
import pt.haslab.dql.tictactoe.util.CommonUtils;

import java.util.Random;

/**
 * Implements a master agent, dubbed MrMiyagi, which follows a hardcoded
 * strategy that makes the best move available at each time.
 * Created by nunomachado on 09/08/17.
 * Adapted from https://github.com/rfeinman/tictactoe-reinforcement-learning/blob/master/agent.py
 */
public class MrMiyagiAgent extends Agent {
    private double proLevel; //probability of following the optimal strategy instead of a random one
    private int mySymbol;

    public MrMiyagiAgent(Seed key, double level)
    {
        proLevel = level;
        mySymbol = key == Seed.CROSS ? 1 : -1;
    }

    /**
     * If we have two in a row and the 3rd is available, take it.
     * @param board
     */
    private String win(int[][] board, int symbol){
        //Check for diagonal wins
        double[] a = {board[0][0], board[1][1], board[2][2]};
        double[] b = {board[0][2], board[1][1], board[2][0]};

        if(CommonUtils.countKey(a,0) == 1 && CommonUtils.countKey(a, symbol) == 2){
            int index = CommonUtils.indexOf(a, 0);
            return Actions.ACTIONS[(3*index+index)];
        }
        else if(CommonUtils.countKey(b,0) == 1 && CommonUtils.countKey(b, symbol) == 2){
            int index = CommonUtils.indexOf(b, 0);
            if(index == 0)
                return Actions.ACTIONS[2];
            else if(index == 1)
                return Actions.ACTIONS[4];
            else
            return Actions.ACTIONS[6];
        }
        //Now check for 2 in a row/column + empty 3rd
        for(int i = 0; i < 3; i++){
            double[] c = {board[0][i], board[1][i], board[2][i]};
            double[] d = {board[i][0], board[i][1], board[i][2]};
            if(CommonUtils.countKey(c,0) == 1 && CommonUtils.countKey(c, symbol) == 2){
                int index = CommonUtils.indexOf(c, 0);
                return Actions.ACTIONS[3*index+i];
            }
            else if(CommonUtils.countKey(d,0) == 1 && CommonUtils.countKey(d, symbol) == 2){
                int index = CommonUtils.indexOf(d, 0);
                return Actions.ACTIONS[3*i+index];
            }
        }
        return null;
    }

    //return the move that will make our opponent win
    private String blockWin(int[][] board){
        return win(board, mySymbol*-1);
    }

    //Create a fork opportunity such that we have 2 threats to win.
    private String fork(int[][] board){
        //Check all adjacent side middles
        if(board[1][0] == mySymbol && board[0][1] == mySymbol){
            if(board[0][0] == 0 && board[2][0] == 0 && board[0][2] == 0)
                return Actions.ACTIONS[0];
            else if(board[1][1] == 0 && board[2][1] == 0 && board[1][2] == 0)
                return Actions.ACTIONS[4];
        }
        else if(board[1][0] == mySymbol && board[2][1] == mySymbol) {
            if (board[2][0] == 0 && board[0][0] == 0 && board[2][2] == 0)
                return Actions.ACTIONS[6];
            else if (board[1][1] == 0 && board[0][1] == 0 && board[1][2] == 0)
                return Actions.ACTIONS[4];
        }
        else if(board[2][1] == mySymbol && board[1][2] == mySymbol) {
            if (board[2][2] == 0 && board[2][0] == 0 && board[0][2] == 0)
                return Actions.ACTIONS[8];
            else if (board[1][1] == 0 && board[1][0] == 0 && board[0][1] == 0)
                return Actions.ACTIONS[4];
        }
        else if(board[1][2] == mySymbol && board[0][1] == mySymbol) {
            if (board[0][2] == 0 && board[0][0] == 0 && board[2][2] == 0)
                return Actions.ACTIONS[2];
            else if (board[1][1] == 0 && board[1][0] == 0 && board[2][1] == 0)
                return Actions.ACTIONS[4];
        }
        //Check all cross corners
        else if(board[0][0] == mySymbol && board[2][2] == mySymbol) {
            if (board[1][0] == 0 && board[2][1] == 0 && board[2][0] == 0)
                return Actions.ACTIONS[6];
            else if (board[0][1] == 0 && board[1][2] == 0 && board[0][2] == 0)
                return Actions.ACTIONS[2];
        }
        else if(board[2][0] == mySymbol && board[0][2] == mySymbol) {
            if (board[2][1] == 0 && board[1][2] == 0 && board[2][2] == 0)
                return Actions.ACTIONS[8];
            else if (board[1][0] == 0 && board[0][1] == 0 && board[0][0] == 0)
                return Actions.ACTIONS[0];
        }
        return null;
    }

    //Block the opponents fork if she has one available.
    private String blockFork(int[][] board){
        double[] corners = {board[0][0], board[2][0], board[0][2], board[2][2]};
        int opponent = mySymbol*-1;
        //Check all adjacent side middles
        if( board[1][0] == opponent && board[0][1] == opponent) {
            if (board[0][0] == 0 && board[2][0] == 0 && board[0][2] == 0)
                return Actions.ACTIONS[0];
            else if (board[1][1] == 0 && board[2][1] == 0 && board[1][2] == 0)
                return Actions.ACTIONS[4];
        }
        else if( board[1][0] == opponent && board[2][1] == opponent) {
            if (board[2][0] == 0 && board[0][0] == 0 && board[2][2] == 0)
                return Actions.ACTIONS[6];
            else if (board[1][1] == 0 && board[0][1] == 0 && board[1][2] == 0)
                return Actions.ACTIONS[4];
        }
        else if( board[2][1] == opponent && board[1][2] == opponent)
            if( board[2][2] == 0 && board[2][0] == 0 && board[0][2] == 0)
                return Actions.ACTIONS[8];
            else if( board[1][1] == 0 && board[1][0] == 0 && board[0][1] == 0)
            return Actions.ACTIONS[4];
        else if( board[1][2] == opponent && board[0][1] == opponent)
            if( board[0][2] == 0 && board[0][0] == 0 && board[2][2] == 0)
                return Actions.ACTIONS[2];
            else if( board[1][1] == 0 && board[1][0] == 0 && board[2][1] == 0)
            return Actions.ACTIONS[4];

        //Check all cross corners (first check for double fork opp using the corners array)
        else if( CommonUtils.countKey(corners,0) == 1 && CommonUtils.countKey(corners,opponent) == 2)
            return Actions.ACTIONS[5];
        else if( board[0][0] == opponent && board[2][2] == opponent)
            if( board[1][0] == 0 && board[2][1] == 0 && board[2][0] == 0)
                return Actions.ACTIONS[6];
            else if( board[0][1] == 0 && board[1][2] == 0 && board[0][2] == 0)
            return Actions.ACTIONS[2];
        else if( board[2][0] == opponent && board[0][2] == opponent)
            if( board[2][1] == 0 && board[1][2] == 0 && board[2][2] == 0)
                return Actions.ACTIONS[8];
            else if( board[1][0] == 0 && board[0][1] == 0 && board[0][0] == 0)
            return Actions.ACTIONS[0];
        return null;
    }

    //Pick the center if it is available.
    public String center(int[][] board){
        if(board[1][1] == 0)
            return Actions.ACTIONS[4];
        return null;
    }

    //Pick a corner move
    public String corner(int[][] board){
        int opponent = mySymbol*-1;
        //Pick opposite corner of opponent if available
        if( board[0][0] == opponent && board[2][2] == 0)
            return Actions.ACTIONS[8];
        else if( board[2][0] == opponent && board[0][2] == 0)
            return Actions.ACTIONS[2];
        else if( board[0][2] == opponent && board[2][0] == 0)
            return Actions.ACTIONS[6];
        else if( board[2][2] == opponent && board[0][0] == 0)
            return Actions.ACTIONS[0];
        //Pick any corner if( no opposites are available
        else if( board[0][0] == 0)
            return Actions.ACTIONS[0];
        else if( board[2][0] == 0)
            return Actions.ACTIONS[6];
        else if( board[0][2] == 0)
            return Actions.ACTIONS[2];
        else if( board[2][2] == 0)
            return Actions.ACTIONS[8];
        return null;
    }

    //Pick an empty side.
    public String sideEmpty(int[][] board){
        if( board[1][0] == 0)
            return Actions.ACTIONS[3];
        else if( board[2][1] == 0)
            return Actions.ACTIONS[7];
        else if( board[1][2] == 0)
            return Actions.ACTIONS[5];
        else if( board[0][1] == 0)
            return Actions.ACTIONS[1];
        return null;
    }

    //agent acts randomly
    public String randomMove(int[][] board) {
        Random r = new Random();
        int a = r.nextInt(Actions.ACTIONS.length);
        return Actions.ACTIONS[a];
    }

    /**
     *  Agent goes through a hierarchy of moves, making the best move that
        is currently available each time.
     * @param state
     * @return
     */
    public String act(Board state) {
        Random r = new Random();
        int[][] board = state.toMatrix();

        //Choose randomly with some probability so that the agent does not always win
        if(r.nextDouble() > proLevel){
            return randomMove(board);
        }
        //Follow optimal strategy
        String a = win(board, mySymbol);
        if(a!=null)
            return a;
        a = blockWin(board);
        if(a!=null)
            return a;
        a = fork(board);
        if(a!=null)
            return a;
        a = blockFork(board);
        if(a!=null)
            return a;
        a = center(board);
        if(a!=null)
            return a;
        a = corner(board);
        if(a!=null)
            return a;
        a = sideEmpty(board);
        if(a!=null)
            return a;
        return randomMove(board);
    }
}

File: tictactoe/agents/mr_miyagi_agent.py
import random

from tictactoe.agents.agent import Agent
from tictactoe.game.board import Board, Seed
from tictactoe.game.moves import Moves


class MrMiyagiAgent(Agent):
    """
    Implements a master agent, dubbed MrMiyagi, which follows a hardcoded
    strategy that makes the best move available at each time.
    Adapted from https://github.com/rfeinman/tictactoe-reinforcement-learning/blob/master/agent.py
    """

    def __init__(self, seed: Seed, level: float):
        self.pro_level = level
        if seed == Seed.CROSS:
            self.seed = 1
        else:
            self.seed = -1

    def find_move(self, board, symbol):
        """Helper function to find winning or blocking move."""
        for i in range(3):
            if board[i].count(symbol) == 2 and board[i].count(0) == 1:
                return Moves.MOVES[i * 3 + board[i].index(0)]
            col = [board[j][i] for j in range(3)]
            if col.count(symbol) == 2 and col.count(0) == 1:
                return Moves.MOVES[col.index(0) * 3 + i]
        else:
            diag1 = [board[i][i] for i in range(3)]
            diag2 = [board[i][2 - i] for i in range(3)]
            if diag1.count(symbol) == 2 and diag1.count(0) == 1:
                index = diag1.index(0)
                return Moves.MOVES[index * 3 + index]
            if diag2.count(symbol) == 2 and diag2.count(0) == 1:
                index = diag2.index(0)
                return Moves.MOVES[index * 3 + (2 - index)]

    def win(self, board):
        """
        Finds a winning move if one exists.
        """
        return self.find_move(board, self.seed)

    def block_win(self, board):
        return self.find_move(board, -self.seed)

    def find_fork(self, board, symbol):
        """Helper function to find fork opportunities."""
        for i, j in [(1, 1), (0, 0), (0, 2), (2, 0), (2, 2)]:
            if board[i][j] == 0:
                board[i][j] = symbol
                moves = sum(
                    (self.find_move(board, symbol) is not None for _ in range(4))
                )
                board[i][j] = 0
                if moves >= 2:
                    return Moves.MOVES[i * 3 + j]
        else:
            return

    def fork(self, board):
        return self.find_fork(board, self.seed)

    def block_fork(self, board):
        return self.find_fork(board, -self.seed)

    def center(self, board):
        if board[1][1] == 0:
            return Moves.MOVES[4]
        return None

    def corner(self, board):
        for i, j in [(0, 0), (0, 2), (2, 0), (2, 2)]:
            if board[i][j] == 0:
                return Moves.MOVES[i * 3 + j]
        else:
            return

    def side_empty(self, board):
        for i, j in [(1, 0), (2, 1), (1, 2), (0, 1)]:
            if board[i][j] == 0:
                return Moves.MOVES[i * 3 + j]
        else:
            return

    def random_move(self):
        return random.choice(Moves.MOVES)

    def make_move(self, state: Board) -> str:
        board = state.to_matrix()
        if random.random() > self.pro_level:
            return self.random_move(board)
        strategies = [
            self.win,
            self.block_win,
            self.fork,
            self.block_fork,
            self.center,
            self.corner,
            self.side_empty,
        ]
        for strategy in strategies:
            action = strategy(board)
            if action:
                return action
        else:
            return self.random_move(board)

    def __str__(self):
        return "MIYAGI"
